### ISSUE TITLE
Closure ser/de error in Scala 2.12

### ISSUE THREAD

**Issue by nevillelyh** [spotify/scio#469]
Happens in Scala 2.12. The following code path in `ClosureCleaner` was executed but didn't clean it properly.
```scala
  case e: IllegalArgumentException if e.getCause.isInstanceOf[NotSerializableException] =>
        com.twitter.chill.ClosureCleaner(obj)
```

```
[info] ClosureTest:
[info] - should support class fn *** FAILED ***
[info]   java.lang.IllegalArgumentException: unable to serialize com.spotify.scio.util.Functions$$anon$7@23b04081
[info]   at org.apache.beam.sdk.util.SerializableUtils.serializeToByteArray(SerializableUtils.java:53)
[info]   at org.apache.beam.sdk.util.SerializableUtils.clone(SerializableUtils.java:90)
[info]   at org.apache.beam.sdk.transforms.ParDo$Bound.<init>(ParDo.java:690)
[info]   at org.apache.beam.sdk.transforms.ParDo$Unbound.of(ParDo.java:664)
[info]   at org.apache.beam.sdk.transforms.ParDo$Unbound.access$000(ParDo.java:596)
[info]   at org.apache.beam.sdk.transforms.ParDo.of(ParDo.java:529)
[info]   at com.spotify.scio.values.PCollectionWrapper.parDo(PCollectionWrapper.scala:55)
[info]   at com.spotify.scio.values.PCollectionWrapper.parDo$(PCollectionWrapper.scala:54)
[info]   at com.spotify.scio.values.SCollectionImpl.parDo(SCollection.scala:1127)
[info]   at com.spotify.scio.values.SCollection.map(SCollection.scala:356)
[info]   ...
[info]   Cause: java.io.NotSerializableException: org.scalatest.Assertions$AssertionsHelper
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
[info]   at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
[info]   at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
[info]   at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
[info]   at java.io.ObjectOutputStream.writeArray(ObjectOutputStream.java:1378)
[info]   at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1174)
[info]   at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
[info]   at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
[info]   at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
```

**Comment by andrewsmartin** [comment]
The `classFn` lambda gets compiled as:

```
public static final org.scalatest.compatible.Assertion $anonfun$new$2(com.spotify.scio.values.ClosureTest, com.spotify.scio.ScioContext);
```

which has an instance of `ClosureTest` (not serializable). `ClosureCleaner` does not seem to handle this case.

**Comment by nevillelyh** [comment]
Related:
https://github.com/twitter/chill/issues/252
https://issues.apache.org/jira/browse/SPARK-14540

**Comment by nevillelyh** [comment]
So #482 fixed the most common `$outer` issue (`$outer` is renamed to `arg$n` in Scala 2.12) but another edge case remains:

```scala
object MyJob {
  def plus(x: Int) = x + 1
  def main(...) = {
    ...
    val p = sc.parallelize(1 to 10)
    p.map(plus)  // #1 this fails
    p.map(MyJob.plus) // #2 this works
    ...
  }
}
```

So it seems `p.map(plus)` now passes `class MyJob$#plus` into `map` and the class has a `MODULE$: MyJob$` field that's not serializable. `p.map(MyJob.plus)` on the other hand works since `MyJob.plus` is a static method on an object.
I dug around `ClosureCleaner` but haven't found a solution yet. Worst case this could be fixed in user land by doing `object MyJob extends Serializable` or passing `MyJob.plus` into transforms but will break code compatibility. Thoughts?

**Comment by nevillelyh** [comment]
Reported to Scala: https://issues.scala-lang.org/browse/SI-10233

**Comment by andrewsmartin** [comment]
I spent some time with `ClosureCleaner` and any fix would involve non-trivial (and probably dangerous) byte-code hacks. It's better to wait for https://github.com/scala/scala/pull/5782 and have the compiler handle this directly.

**Comment by nevillelyh** [comment]
Fixed in scala 2.12.3
